File: src/leetCode/0_350_IntersectionOfTwoArraysII/hash.ts
function intersectHash(nums1: number[], nums2: number[]): number[] {
  if (nums1.length > nums2.length) {
    return intersectHash(nums2, nums1)
  }

  const results: number[] = [];
  const hash = new Map<number, number>()

  nums1.forEach(n1 => {
    if (!hash.has(n1)) {
      hash.set(n1, 1)

      return
    }

    hash.set(n1, hash.get(n1) + 1)
  })

  nums2.forEach(n2 => {
    if (!hash.has(n2)) {
      return
    }

    const times = hash.get(n2)

    if (times > 0) {
      hash.set(n2, times - 1)
      results.push(n2)
    }
  })

  return results;
};

File: src/leetCode/0_350_IntersectionOfTwoArraysII/hash.go
package intersect

// https://leetcode-cn.com/problems/intersection-of-two-arrays-ii/
// https://www.geekxh.com/1.0.%E6%95%B0%E7%BB%84%E7%B3%BB%E5%88%97/001.html
// HASH表
func intersectHash(nums1 []int, nums2 []int) []int {
	// 把小数组写入HASH省空间
	if len(nums1) > len(nums2) {
		return intersectHash(nums2, nums1)
	}

	hash := make(map[int]int)
	results := make([]int, 0)

	for _, num1 := range nums1 {
		hash[num1] += 1
	}

	for _, num2 := range nums2 {
		if hash[num2] > 0 {
			hash[num2] -= 1

			results = append(results, num2)
		}
	}

	return results
}
